accepted/2023/sign-signature-provider-plugins.md

```diff
@@ -0,0 +1,524 @@
+
+#### Package versioning
+
+The interfaces package version will follow strict [SemVer 2.0.0](https://semver.org/spec/v2.0.0.html) versioning rules.  It is expected that no interfaces package version will introduce a breaking change over any previous version.  Because the interfaces package is intended to provide stable abstractions to both Sign CLI and plugins, every package version will be fully backwards compatible.  Therefore, it is expected that we will only release packages in the version range [1.0.0, 2.0.0).  If a breaking change is warranted, versions >= 2.0.0 are a possibility.
+
+It is intended that we will only publish release versions of the interfaces package.  If, for any reason, we decide to publish prerelease versions later, it should still be assumed that official Sign CLI releases (even prerelease versions) will only reference release versions of the interfaces package.  For any given official Sign CLI release, Sign CLI should reference the latest release version of the interfaces package at that point in time.
+
+#### Interface versioning
+
+Interfaces defined in the interfaces package should be considered permanent and immutable.  New interfaces can be added, but existing interfaces should not be modified or removed as long as Sign CLI will load plugins that implement those existing interfaces.
+
+Versioning strategies for plugin interfaces are out of scope for this specification, but it is expected that all interfaces for every supported version of the plugin will be available in the interfaces package to enable Sign CLI users to install not only the latest version but previous, supported versions as well.
+
+> Note:  If a plugin author wanted to remove support for older versions of a plugin, we could achieve that by specifying a minimum version of the plugin package in Sign CLI.  Older versions would be ignored, and plugin interfaces for those older versions could safely be removed from the interfaces package, provided that they are no longer needed.  Then, it would be possible for a plugin author to remove support for older versions and modify existing interfaces (vs. add new interfaces) in one step.  This remains true to the spirit of earlier guidance, that the interfaces package should preserve interfaces for all _supported_ plugin versions.  Enabling plugin authors to drop support for older versions of a plugin is out of scope for this specification.
+
+### Plugins
+
+A Sign CLI signature provider plugin:
+
+* extends Sign CLI functionality
+* contains a [`plugin.json`](#plugin-json-file) file in its root directory
```

Comment:
I'm worried that this might lead to conflicts if people look at this work as a reference implementation. We might end up with a model where lots of .NET apps require a plugin.json for their plugins. I can imagine a case where some plugin can be a plugin for multiple apps.

Reply:
I would agree - would it be better to base the plugin discovery on patterns?
* Name of the file (`Microsoft.Azure.KeyVault.SignPlugin.dll` -> `*.SignPlugin.dll`) ?
* Use `System.Reflection.Metadata` to load the assembly (but not for running) and inspect it - from there you could read the necessary metadata (name, description, ...) - one option would be assembly level attributes for example (just like version information is stored today).

Reply:
@richlander, @vitek-karas, I understand your concerns.  We drew inspiration from dotnet/templating's template.json file as a reference on how to declare assets and contributions.  Here, plugin.json provides 2 things:  a mechanism for locating a plugin's main assembly and CLI option metadata.

I'm not sure how to entirely eliminate your concerns without tackling a generic .NET tool plugin model at the same time, such that dotnet/sign follows the same pattern.  A general .NET tool plugin model wasn't originally in scope, but we're open to working with others to find out how we can deliver this in alignment with a generic .NET tool plugin model.  There is pressure for dotnet/sign to deliver something in the very near future.

We could use an assembly file name pattern instead and remove that section from plugin.json.  plugin.json still specifies CLI option metadata, and we wanted to leave the door open for localization, even though it's out of scope now.   @clairernovotny and I talked about using attributes in the assembly to provide CLI option metadata, potentially generating plugin.json at build time (similar to dotnet/templating), but that all felt nice-to-have not MVP.

Are you suggesting that we eliminate plugin.json entirely or that a plugin.json file should conform to a standard .NET tool plugin schema?